pandas/core/arrays/base.py

```diff
@@ -486,10 +490,11 @@ def to_numpy(
         -------
         numpy.ndarray
         """
-        result = np.asarray(self, dtype=dtype)
```

Comment:
wouldn't it be simpler to just call `result = np.array(self, dtype=dtype, copy=copy)`?

Reply:
@guitargeek gentle ping would this work?

Reply:
No, this would not work I think. The problem is that there is no way to know if `__array__` has returned a copy. Just like before my PR proposal, `np.array` would always do a copy if if `copy=True` here, even if `__array__` actually already returned a copy. That's again the original double-copy problem that the PR was opened to address.

I currently don't see any way around introducing a new function like `_to_numpy`, which is behaving almost like `__array__` only that it returns a second flag to tell if a copy was already made.

Reply:
I must be missing something here, please bear with me.  I don't see why we need to know whether `__array__` made a copy.

Suppose on L489 we did `result = np.array(self, dtype=dtype, copy=copy)`.  If the user passed `copy=True`, then we know that `result` is a copy, and now know we don't need to make any subsequent copies.  If the user passed `copy=False`, we don't know if `result` is a copy or not, but we _still_ know we don't want to make any subsequent copies.

Reply:
The problem is that the double copy already happens inside `result = np.array(self, dtype=dtype, copy=copy)` here. As I understand, `np.array` calls `__array__`, and if `copy=True` is will make a copy of what `__array__` returned. If `__array__` already returns a copy, as is the case for many custom pandas array classes, we get a double copy.

Reply:
I see.  I was under the mistaken impression that `copy` was passed to `__array__`

@seberg is there a nicer way to check if `np.array(obj, copy=False)` made a copy for an arbitrary `obj`?

Reply:
No, I proposed adding a `ensure_view=True` flag or so, but we would have to pass information out of `__array__` and we cannot without extending the protocol.  None of the protocols really provides the information, some are "probably not a copy" (`__array__` of course is not one of those).
I.e. we need a proposal to extend `__array__` to really get it right.  I had hoped when the next protocol comes around, it will take this into account, but right now I think extending `__array__` may well be the best bet we got, although I am not sure exactly how :).